### ISSUE TITLE
floating-point models very dependent on platform

### ISSUE THREAD

**Issue by yannickmoy** [CVC4/CVC4#2285]
We are now exploiting the floating-point models returned by CVC4, and we've noticed that the models are completely different between Mac OS and other platforms (Linux 32bits and 64bits, Windows 32bits). For example on the VC:

(set-info :smt-lib-version 2.5)
(set-logic AUFBVFPDTNIRA)
(set-info :source |VC generated by SPARK 2014|)
(set-info :category industrial)
(set-info :status unknown)
(set-option :produce-models true)

(define-fun fp.isFinite32 ((x Float32)) Bool (not (or (fp.isInfinite x) (fp.isNaN x))))
(declare-const a Float32)
(declare-const b Float32)
(assert (fp.isFinite32 a))
(assert (fp.isFinite32 b))
(assert (fp.isFinite32 (fp.add RNE a b)))

(check-sat)
(get-model)

CVC4 generates the following model on all platforms:

(model
(define-fun a () (_ FloatingPoint 8 24) (fp #b0 #b01111111 #b00000000000000000000010))
(define-fun b () (_ FloatingPoint 8 24) (fp #b0 #b01111111 #b00000000100000000000000))
)

except on Mac OS on which it generates:

(model
(define-fun a () (_ FloatingPoint 8 24) (fp #b0 #b01111111 #b10000010000000000000010))
(define-fun b () (_ FloatingPoint 8 24) (fp #b1 #b01111111 #b01000010000000000000000))
)

This is a behavior we've not seen previously for models on other domains. Typically most models end up being the same across platforms, with a few differences sometimes.

First, I'd like to check with you that it's expected, and not the result of a hidden bug.

Secondly, it's cumbersome to maintain platform-dependent outputs for a large number of tests, so would it be possible to avoid this difference for Mac OS? Can we use a switch, or a different way of building CVC4, to avoid that difference?

We are currently following CVC4 master commit bc9365678

**Comment by martin-cs** [comment]
On Wed, 2018-08-08 at 13:47 +0000, yannickmoy wrote:
> We are now exploiting the floating-point models returned by CVC4

Great to hear!


> This is a behavior we've not seen previously for models on other
> domains. Typically most models end up being the same across
> platforms, with a few differences sometimes.
> 
> First, I'd like to check with you that it's expected, and not the
> result of a hidden bug.

To my knowledge, CVC4 guarantees to be deterministic but nothing else
about the stability of the models produced ( @barrettcw please correct
me if I am wrong ).  I haven't seen this before but I am not entirely
surprised and IMHO it's not a bug as long as both of the models are
actually satisfying assignments.


> Secondly, it's cumbersome to maintain platform-dependent outputs for
> a large number of tests, so would it be possible to avoid this
> difference for Mac OS? Can we use a switch, or a different way of
> building CVC4, to avoid that difference?

I am not aware of any such flag or build mode.  As for tracking down
the discrepancy; my guess would be that it is a piece of platform
dependent behaviour in one of the libraries.  Something like iterating
across a hash_map.  BUT... there are an awfully large number of places
that could be.  Diff'ing a full set of trace outputs might help but TBH
I think it's going to be a big job and bit-identical results on all
platforms, unless it becomes a design aim and we CI for it, is going to
be challenging.

Also note that which satisfying assignment you get is only
deterministic for one version of the solver.  They may change per
commit, so even if they are the same on each platform, you are still
going to get changes as you update.

**Comment by yannickmoy** [comment]
As far as it's expected behavior for you, we'll work around it in our testsuite, thanks!
Changes in behavior when we update CVC4 are expected and we are fine with that.

**Comment by barrettcw** [comment]
I agree with Martin.  We aim to make CVC4 deterministic across runs on the
same platform with the same binary, but beyond that, there are no
guarantees.

It is actually surprising that you have seen consistent results across
platforms up until now which suggests that maybe the problem could be fixed
in the short term.  But, as Martin said, figuring out the reason could be
challenging and anyway, making this guarantee in the long term is not
something we are set up to support.

-Clark

**Comment by martin-cs** [comment]
Perhaps I was being too negative.  I believe it to be technically
feasible.  I think the initial exploratory work and CI infrastructure
to catch issues to be about 3 person-months.  On top of that, there
will likely be an additional amount of time to actually solve the
problem.  At the moment I don't have that time to spare.  I can't speak
for other members of the CVC4 team but like all successful projects
(and I'm sure you can sympathise w.r.t. SPARK) there is a general sense
of 'more things to do than time to do it'.  However whoever did it
would undoubtably learn a lot about the internals of CVC4.

So @yannickmoy how useful would it be to you?  Could you help raise the
money for someone to do it?

**Comment by yannickmoy** [comment]
Good to know that this could be doable depending on funding! Let me consider it for 2019 budget and I'll come back to you.
